# Linear Models with SparkR: uses and present limitations

[**Introduction to Apache Spark with R by J. A. Dianes**](https://github.com/jadianes/spark-r-notebooks)

In this notebook we will use SparkR machine learning capabilities in order to predict property value in relation to other variables in the [2013 American Community Survey](http://www.census.gov/programs-surveys/acs/data/summary-file.html) dataset. The whole point of R on Spark is to introduce Spark scalability into R data analysis pipelines. With this idea in mind, we have seen how [SparkR documentation](http://spark.apache.org/docs/latest/sparkr.html) introduces data types and functions that are very similar to what we are used to when using regular R. The nest step in our series of notebooks will deal with its machine learning capabilities. While building a linear model we will also check the significance of each of the variables involved in building such a predictor for property value.

## Creating a SparkSQL context and loading data

In order to explore our data, we first need to load it into a SparkSQL data frame. But first we need to init a SparkSQL context. The first thing we need to do is to set up some environment variables and library paths as follows. Remember to replace the value assigned to `SPARK_HOME` with your Spark home folder.  

In [1]:
# Set Spark home and R libs
Sys.setenv(SPARK_HOME='/home/cluster/spark-1.5.0-bin-hadoop2.6')
.libPaths(c(file.path(Sys.getenv('SPARK_HOME'), 'R', 'lib'), .libPaths()))

Now we can load the `SparkR` library as follows.

In [2]:
library(SparkR)


Attaching package: ‘SparkR’

The following objects are masked from ‘package:stats’:

    filter, na.omit

The following objects are masked from ‘package:base’:

    intersect, rbind, sample, subset, summary, table, transform



And now we can initialise the Spark context as [in the official documentation](http://spark.apache.org/docs/latest/sparkr.html#starting-up-sparkcontext-sqlcontext). In our case we are use a standalone Spark cluster with one master and seven workers. If you are running Spark in local node, use just `master='local'`. Additionally, we require a Spark package from Databricks to read CSV files (more on this in the [previous notebook](https://github.com/jadianes/spark-r-notebooks/blob/master/notebooks/nb1-spark-sql-basics/nb1-spark-sql-basics.ipynb)). 

In [3]:
sc <- sparkR.init(master='spark://169.254.206.2:7077', sparkPackages="com.databricks:spark-csv_2.11:1.2.0")

Launching java with spark-submit command /home/cluster/spark-1.5.0-bin-hadoop2.6/bin/spark-submit  --packages com.databricks:spark-csv_2.11:1.2.0 sparkr-shell /tmp/RtmpmF2Hmf/backend_port6c9817062e87 


And finally we can start the SparkSQL context as follows.

In [4]:
sqlContext <- sparkRSQL.init(sc)

Now that we have our SparkSQL context ready, we can use it to load our CSV data into data frames. We have downloaded our [2013 American Community Survey dataset](http://www.census.gov/programs-surveys/acs/data/summary-file.html) files in [notebook 0](https://github.com/jadianes/spark-r-notebooks/tree/master/notebooks/nb0-starting-up/nb0-starting-up.ipynb), so they should be stored locally. Remember to set the right path for your data files in the first line, ours is `/nfs/data/2013-acs/ss13husa.csv`.  

In [5]:
housing_a_file_path <- file.path('', 'nfs','data','2013-acs','ss13husa.csv')
housing_b_file_path <- file.path('', 'nfs','data','2013-acs','ss13husb.csv')

Now let's read into a SparkSQL dataframe. We need to pass four parameters in addition to the `sqlContext`:  

- The file path.  
- `header='true'` since our `csv` files have a header with the column names. 
- Indicate that we want the library to infer the schema.  
- And the source type (the Databricks package in this case). 

And we have two separate files for both, housing and population data. We need to join them.

In [6]:
housing_a_df <- read.df(sqlContext, 
                        housing_a_file_path, 
                        header='true', 
                        source = "com.databricks.spark.csv", 
                        inferSchema='true')

In [7]:
housing_b_df <- read.df(sqlContext, 
                        housing_b_file_path, 
                        header='true', 
                        source = "com.databricks.spark.csv", 
                        inferSchema='true')

In [8]:
housing_df <- rbind(housing_a_df, housing_b_df)

Let's check that we have everything there by counting the files and listing a few of them.

In [9]:
nrows <- nrow(housing_df)
nrows

[1] 1476313

In [10]:
head(housing_df)

RT SERIALNO DIVISION PUMA REGION ST  ADJHSG  ADJINC WGTP NP TYPE ACCESS ACR
1  H       84        6 2600      3  1 1000000 1007549    0  1    3     NA  NA
2  H      154        6 2500      3  1 1000000 1007549   51  4    1      1   1
3  H      156        6 1700      3  1 1000000 1007549  449  1    1      3   1
4  H      160        6 2200      3  1 1000000 1007549   16  3    1      3   3
5  H      231        6 2400      3  1 1000000 1007549   52  1    1      3  NA
6  H      286        6  900      3  1 1000000 1007549   76  1    1      1   1
  AGS BATH BDSP BLD BROADBND BUS COMPOTHX CONP DIALUP DSL ELEP FIBEROP FS FULP
1  NA   NA   NA  NA       NA  NA       NA   NA     NA  NA   NA      NA  2   NA
2  NA    1    3   2        2   2        2   NA      2   2  350       1  2    2
3  NA    1    3   2       NA   2        2   NA     NA  NA  300      NA  2    2
4   1    1    4   2       NA   2        2   NA     NA  NA  220      NA  1    2
5  NA    1    1   5       NA  NA        2   NA     NA  NA   60      NA  1    1
6  NA    1    4   2        2   2        2   NA      2   2  100       2  2    2
  GASP HANDHELD HFL INSP LAPTOP MHP MODEM MRGI MRGP MRGT MRGX OTHSVCEX REFR
1   NA       NA  NA   NA     NA  NA    NA   NA   NA   NA   NA       NA   NA
2    3        2   3  350      1  NA     2   NA   NA   NA    3        2    1
3    3        2   3  980      1  NA    NA    1  550    2    1       NA    1
4   20        1   3   NA      2  NA    NA   NA   NA   NA   NA       NA    1
5    3        2   1   NA      2  NA    NA   NA   NA   NA   NA       NA    1
6   90        2   1   50      2  NA     1    2  290    2    1        2    1
  RMSP RNTM RNTP RWAT RWATPR SATELLITE SINK SMP STOV TEL TEN TOIL VACS  VALP
1   NA   NA   NA   NA     NA        NA   NA  NA   NA  NA  NA   NA   NA    NA
2    9   NA   NA    1      9         2    1  NA    1   1   2    1   NA 25000
3    6   NA   NA    1      9        NA    1  NA    1   1   1    1   NA 80000
4    6    2  100    1      9        NA    1  NA    1   1   3    1   NA    NA
5    3    2   80    1      9        NA    1  NA    1   1   3    1   NA    NA
6    9   NA   NA    1      9         2    1  NA    1   1   1    1   NA 18000
  VEH WATP YBL FES FFINCP FGRNTP FHINCP  FINCP FPARC FSMOCP GRNTP GRPIP HHL HHT
1  NA   NA  NA  NA     NA     NA     NA     NA    NA     NA    NA    NA  NA  NA
2   3  480   2   1      1     NA      1 151000     4      0    NA    NA   1   1
3   1  700   5  NA      0     NA      1     NA    NA      1    NA    NA   1   6
4   2  360   6   8      1      0      1  11400     2     NA   370    39   1   3
5   0    1   5  NA      0      0      0     NA    NA     NA   140    43   1   6
6   1  370   2  NA      0     NA      1     NA    NA      1    NA    NA   1   6
   HINCP HUGCL HUPAC HUPAOC HUPARC KIT LNGI MULTG MV NOC NPF NPP NR NRC OCPIP
1     NA    NA    NA     NA     NA  NA   NA    NA NA  NA  NA  NA NA  NA    NA
2 151000     0     4      4      4   1    1     1  6   0   4   0  0   0     3
3  39930     0     4      4      4   1    1     1  7   0  NA   0  0   0    28
4  11400     1     2      4      2   1    1     1  4   0   3   1  0   1    NA
5   3900     0     4      4      4   1    1     1  6   0  NA   0  0   0    NA
6   5400     0     4      4      4   1    1     1  7   0  NA   0  0   0   101
  PARTNER PLM PSF R18 R60 R65 RESMODE SMOCP SMX SRNT SSMC SVAL TAXP WIF WKEXREL
1      NA  NA  NA  NA  NA  NA      NA    NA  NA   NA   NA   NA   NA  NA      NA
2       0   1   0   0   0   0       2   426  NA    0    0    1    3   2       1
3       0   1   0   0   1   0       2   926   3    0    0    1    6  NA      NA
4       0   1   0   1   1   0       2    NA  NA    0    0    0   NA   1      15
5       0   1   0   0   1   1       1    NA  NA    1    0    0   NA  NA      NA
6       0   1   0   0   1   1       1   522   3    0    0    1    3  NA      NA
  WORKSTAT FACCESSP FACRP FAGSP FBATHP FBDSP FBLDP FBROADBNDP FBUSP FCOMPOTHXP
1       NA       NA    NA    NA     NA    NA    NA         NA    NA         NA
2       

## Preparing our data

We need to convert `ST` (or any other categorical variable) from a numeric variable into a factor.

In [11]:
housing_df$ST <- cast(housing_df$ST, "string")

In [12]:
housing_df$REGION <- cast(housing_df$REGION, "string")

Additionally, we need either to impute values or to remove samples with null values in any of our predictors or desponse. For the response (`VALP`) we will use just those samples with actual values.

In [13]:
housing_with_valp_df <- filter(
    housing_df, 
    isNotNull(housing_df$VALP) 
    & isNotNull(housing_df$TAXP)
    & isNotNull(housing_df$INSP)
    & isNotNull(housing_df$ACR)
)

Let's count the remaining samples.

In [14]:
nrows <- nrow(housing_with_valp_df)
nrows

[1] 807202

## Preparing a train / test data split

We don't have a split function in SparkR, but we can use `sample` in combination with the `SERIALNO` column in order to prepare two sets of IDs for training and testing.

In [15]:
housing_df_test <- sample(housing_with_valp_df,FALSE,0.1)

In [16]:
nrow(housing_df_test)

[1] 80511

In [17]:
test_ids <- collect(select(housing_df_test, "SERIALNO"))$SERIALNO

Unfortunately SparkR doesn't support negative %in% expressions, so we need to do this in two steps. First we add a flag to the whole dataset indicating that a sample belongs to the test set.

In [18]:
housing_with_valp_df$IS_TEST <- housing_with_valp_df$SERIALNO %in% test_ids

And then we use that flag to subset out the train dataset as follows.

In [19]:
housing_df_train <- subset(housing_with_valp_df, housing_with_valp_df$IS_TEST==FALSE)

In [20]:
nrow(housing_df_train)

[1] 726691

However this approach is not very scalable since we are collecting all the test IDs and passing them over to build the new flag column. What if we have a much larger test set? Hopefully futre versions of SparkR will come up with a proper `split` functionality.

## Training a linear model

In order to train a linear model, we call `glm` with the following parameters:
- A formula: sadly, `SparkR::glm()` gives us an error when we pass more than eight variables using `+` in the formula.
- The dataset we want to use to train the model.
- The type of model (gaussian or binomial).  

This doesn't differ much from the usual R `glm` command, although right now is more limited.

The list of variables we have used includes:  

- `RMSP` or number of rooms.
- `ACR` the lot size.
- `INSP` or insurance cost.
- `TAXP` or taxes cost.
- `ELEP` or electricity cost.
- `GASP` or gas cost.
- `ST` that is the state code.
- `REGION` that identifies the region.

In [21]:
model <- glm(
    VALP ~ RMSP + ACR + INSP + TAXP + ELEP + GASP + ST + REGION, 
    data = housing_df_train, 
    family = "gaussian")

In [22]:
summary(model, signif.stars=TRUE)

$coefficients
                 Estimate
(Intercept)  -76528.93436
RMSP          14429.66764
ACR           31462.28079
INSP             99.35929
TAXP           5625.89535
ELEP            194.35718
GASP            233.60859
ST__6         83708.07349
ST__48      -422438.29030
ST__12      -373593.91727
ST__36      -162774.32734
ST__42      -187752.54696
ST__39       -92930.33601
ST__17      -104259.08569
ST__26       -94614.73430
ST__37      -331767.13022
ST__13      -361383.26052
ST__51      -272700.96543
ST__34      -174525.27108
ST__18       -41841.98399
ST__47      -337911.59808
ST__53       -80898.74154
ST__55      -104367.12002
ST__29       -66378.45165
ST__27       -79352.27547
ST__4        -63673.30614
ST__24      -294930.50627
ST__25       -96789.48028
ST__1       -314002.67988
ST__8        -60126.28786
ST__45      -314901.99064
ST__21      -348317.83261
ST__22      -330038.12498
ST__41       -99094.19158
ST__40      -374563.06588
ST__19       -82441.48692
ST__9       -140932.88092
ST__5       -337026.43343
ST__20      -111871.58753
ST__28      -352687.68069
ST__49       -75785.25741
ST__32       -90557.15684
ST__54      -310321.26380
ST__35       -65105.29797
ST__31      -124406.90034
ST__16       -95069.84672
ST__23      -145797.09608
ST__33      -216322.49983
ST__30       -99628.00723
ST__10      -244316.55593
ST__44      -183127.44253
ST__15       264858.91571
ST__46       -69061.88061
ST__38       -63042.22881
ST__50      -208646.37622
ST__56      -103393.59207
ST__2        -65145.47470
REGION__3    202606.89664
REGION__2   -113094.23222
REGION__4     -3672.89457

Sadly, the current version of `SparkR::summary()` doesn't provide significance starts. That makes model interpretation and selection very difficult. But at least we know how each variables influences a property value. For example, the Midwest region decreases property value, while the West increases it, etc. In order to interpret that we need to have a look at our [data dictionary](http://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMSDataDict13.txt).

In any case, since we don't have significance starts, we can iterate through adding/removing variables and calculating the R2 value. In our case we ended up with the previous model.

## Evaluating our model using the test data

First of all let's obtain the average value for `VALP` that we will use as a reference of a base predictor model.

In [23]:
VALP_mean <- collect(agg(
    housing_df_train, 
    AVG_VALP=mean(housing_df_train$VALP)
))$AVG_VALP

In [24]:
VALP_mean

[1] 245616.5

Let's now predict on our test dataset as follows.

In [25]:
predictions <- predict(model, newData = housing_df_test)

Let's add the squared residuals and squared totals so later on we can calculate [R2](https://en.wikipedia.org/wiki/Coefficient_of_determination).

In [26]:
predictions <- transform(
    predictions, 
    S_res=(predictions$VALP - predictions$prediction)**2, 
    S_tot=(predictions$VALP - VALP_mean)**2)
head(select(predictions, "VALP", "prediction", "S_res", "S_tot"))

VALP prediction       S_res        S_tot
1  18000   35710.71   313669413  51809288518
2  60000  114905.23  3014584491  34453499299
3 750000  861633.82 12462110173 254402676414
4 300000  196937.20 10621941691   2957560904
5  40000   22703.85   299156942  42278160832
6  60000  121898.49  3831423241  34453499299

In [27]:
nrows_test <- nrow(housing_df_test)
residuals <- collect(agg(
    predictions, 
    SS_res=sum(predictions$S_res),
    SS_tot=sum(predictions$S_tot)
))

In [28]:
residuals

SS_res       SS_tot
1 5.108513e+15 8.635319e+15

In [29]:
R2 <- 1.0 - (residuals$SS_res/residuals$SS_tot)

In [30]:
R2

[1] 0.4084165

In regression, the R2 coefficient of determination is a statistical measure of how well the regression line approximates the real data points. An R2 of 1 indicates that the regression line perfectly fits the data.

A value of 0.41 doesn't speak very well about our model.

## Conclusions

We still need to improve our model if we really want to be able to predict property values. However there are some limitations in the current SparkR implementation that stop us from doing so. Hopefully these limitations won't be there in further versions. Moreover, we are using a linear model, and the relationships between our predictors and the target variable might not be linear at all.

But right now, in Spark v1.5, the R machine learning capabilities are still very limited. We are missing a few things, such as: 

- Accepting more than 8 variables in formulas using `+`.  
- Having significance stars that help model interpretation and selection.  
- Having other indicators (e.g. R2) in summary objects so we don't have to calculate them ourselves.
- Being able to create more complex formulas (e.g. removing intercepts using 100 + ...) so we don't get negative values, etc.
- Although we have a `sample` method, we are missing a `split` one that we can use to easier have train/test splits.  
- Being able to use more powerful models (or at least models that deal better with non linearities), and not just linear ones.